tests/integration/sandbox.py

```diff
@@ -762,6 +770,216 @@ def restart(self):
+    def __init__(self, elle: Elle, clusters: typing.List[Cluster], keys: typing.Optional[List[str]] = None):
+        super().__init__()
+
+        self.clusters = clusters
+        self.elle: Elle = elle
+        self.ended: bool = False
+        self.id: int = 0
+        self.keys: typing.List
+        if keys is not None:
+            self.keys = keys
+        else:
+            self.keys = ["test"]
+
+    def finish(self):
+        self.ended = True
+
+    def set_keys(self, keys: typing.List):
+        self.keys = keys
+
+    # might not want generate_ops, perhaps each multi should just be a single append/read of same key
+    def generate_ops(self, available_keys: typing.List[str]) -> typing.List[ElleOp]:
+        ops: typing.List[ElleOp] = []
+        for _ in range(random.randint(1, 1)):
+            key = available_keys[random.randrange(0, len(available_keys))]
+            val = self.elle.get_next_value()
+            ops.append(ElleOp("append", key, str(val)))
+            ops.append(ElleOp("read", key))
+
+        return ops
+
+    def do_ops(self, ops: typing.List[ElleOp]):
+        # needs a way to handle asking
+        good_write = False
+        needs_asking = False
+
+        client = redis.Redis(host="localhost", port=self.clusters[0].leader_node().port)
+        self.elle.log_command(self.id, ops)
+
+        # 3 possible hops in normal scenario
+        # default MOVED -> ASK -> (remote) MOVED (not leader, so ASKING again) -> "remoe leader"
+        # if it doesn't work by then, error
+        conn: typing.Optional[RawConnection] = None
+        for i in range(0, 3):
+            if conn is not None:
+                conn.release()
+
+            conn = RawConnection(client)
+
+            if needs_asking:
+                assert conn.execute('ASKING') == b'OK'
+
+            assert conn.execute('MULTI') == b'OK'
+
+            # queue up operations
+            for j in range(len(ops)):
+                if ops[j].op_type == "append":
+                    assert conn.execute('RPUSH', ops[j].key, ops[j].value) == b'QUEUED'
+                elif ops[j].op_type == "read":
+                    assert conn.execute('LRANgE', ops[j].key, 0, -1) == b'QUEUED'
+                else:
+                    raise Exception(f"Unknown op_type {ops[j].op_type}")
+
+            # exec/handle failure
+            try:
+                ret = conn.execute('EXEC')
+                self.elle.log_result(self.id, "ok", ret, ops)
+                good_write = True
+                break
+            except ResponseError as e:
+                # handle MOVED/ASK as those aren't "failures"
+                m = re.search(r"^MOVED \d* (.*):(.*)$", str(e))
+                a = re.search(r"^ASK \d* (.*):(.*)$", str(e))
+                if m:
+                    client = redis.Redis(host=m.group(1), port=m.group(2))
+                elif a:
+                    # don't know how to handle submitting an ASKING yet + how to control it
```

Comment:
Is this comment still valid? Looks like you handle asking already.

Reply:
yes, handle it now, need to remove